pkg/networkpath/traceroute/runner.go

```diff
@@ -135,45 +171,106 @@ func (r *Runner) RunTraceroute(ctx context.Context, cfg Config) (payload.Network
 	}
 
-	hname, err := hostname.Get(ctx)
+	pathResult, err := r.processUDPResults(results, hname, cfg.DestHostname, destPort, dest)
 	if err != nil {
 		return payload.NetworkPath{}, err
 	}
+	log.Debugf("Processed Results: %+v", results)
+
+	return pathResult, nil
+}
+
+func (r *Runner) runTCP(cfg Config, hname string, target net.IP, maxTTL uint8, timeout time.Duration) (payload.NetworkPath, error) {
+	destPort := cfg.DestPort
+	if destPort == 0 {
+		destPort = 80 // TODO: is this the default we want?
+	}
 
-	pathResult, err := r.processResults(results, hname, rawDest, destPort, dest)
+	tr := tcp.TCPv4{
+		Target:   target,
+		DestPort: destPort,
+		NumPaths: 1,
+		MinTTL:   uint8(DefaultMinTTL),
+		MaxTTL:   maxTTL,
+		Delay:    time.Duration(DefaultDelay) * time.Millisecond,
+		Timeout:  timeout,
+	}
+
+	results, err := tr.TracerouteSequential()
 	if err != nil {
 		return payload.NetworkPath{}, err
 	}
-	log.Debugf("Processed Results: %+v", results)
+	log.Debugf("TCP Results: %+v", results)
+
+	pathResult, err := r.processTCPResults(results, hname, cfg.DestHostname, destPort, target)
+	if err != nil {
+		return payload.NetworkPath{}, err
+	}
+	log.Debugf("Parsed TCP Results: %+v", pathResult)
 
-	// TODO: better tagging
-	tracerouteRunnerTelemetry.runs.Inc()
 	return pathResult, nil
 }
 
-func getPorts(configDestPort uint16) (uint16, uint16, bool) {
-	var destPort uint16
-	var srcPort uint16
-	var useSourcePort bool
-	if configDestPort > 0 {
-		// Fixed Destination Port
-		destPort = configDestPort
-		useSourcePort = true
-	} else {
-		// Random Destination Port
-		destPort = DefaultDestPort + uint16(rand.Intn(30))
-		useSourcePort = false
+func (r *Runner) processTCPResults(res *tcp.Results, hname string, destinationHost string, destinationPort uint16, destinationIP net.IP) (payload.NetworkPath, error) {
+	pathID := uuid.New().String()
+	traceroutePath := payload.NetworkPath{
+		PathID:    pathID,
+		Timestamp: time.Now().UnixMilli(),
+		Source: payload.NetworkPathSource{
+			Hostname:  hname,
+			NetworkID: r.networkID,
+		},
+		Destination: payload.NetworkPathDestination{
+			Hostname:  getDestinationHostname(destinationHost),
+			Port:      destinationPort,
+			IPAddress: destinationIP.String(),
+		},
 	}
-	srcPort = DefaultSourcePort + uint16(rand.Intn(10000))
-	return destPort, srcPort, useSourcePort
+
+	// get hardware interface info
+	//
+	// TODO: using a gateway lookup may be a more performant
+	// solution for getting the local addr to use
+	// when sending traceroute packets in the TCP implementation
+	// really we just need the router piece
+	// might be worth also looking in to sharing a router between
+	// the gateway lookup and here or exposing a local IP lookup
+	// function
+	if r.gatewayLookup != nil {
+		src := util.AddressFromNetIP(res.Source)
+		dst := util.AddressFromNetIP(res.Target)
+
+		traceroutePath.Source.Via = r.gatewayLookup.LookupWithIPs(src, dst, r.nsIno)
```

Comment:
I feel like this is fine to start, but NPM should already be doing a gateway lookup per IP -- have you thought about ways to share the lookup ?

Reply:
Yes, when I modified the Gateway lookup to work with traceroute, I tried to see if there was a way to share it and tried for a while to make it work but had some difficulties. I discussed with @hmahmood at the time and we agreed for now it was best to keep them separate.

I agree though, it would definitely be better if we could create some sort of singleton to share the caches or something along those lines.